### PR TITLE
Feature/gh 116 support for copy endpoints

### LINKED ISSUE #116: Support for copy endpoints and payloads
Support for copy endpoints and payloads

### PR DESCRIPTION
**Pull request by OButlerOcc**
Support PullRequest adding the copy endpoints and payload models for dataModel/DataElement/DataClass, Terminology/Term and CodeSet

Supports: https://github.com/MauroDataMapper/mdm-ui/compare/feature-gh-861-copy-items-ui?expand=1

closes #116

### CODE REVIEW

File: src/data-models/mdm-data-class.resource.ts [hunk 1]
```diff
@@ -499,4 +499,19 @@ export class MdmDataClassResource extends MdmResource {
...
+  /*
+    * `HTTP POST` - Copy a data class from a source data model to a target data model.
+    *
+    * @param CopyDataClassPyaload The payload of the request containing the source and target data model and data class IDs.
+    * @returns The result of the `POST` request.
+    *
+    * `200 OK` - will return a {@link DataClassDetailResponse} containing a {@link DataClassDetail} object.
+  */
+
+  copy(dataClassId: Uuid, copyDataClassPayload: CopyDataClassPayload){
```
Comment (pjmonks) [inline src/data-models/mdm-data-class.resource.ts]: I think it's interesting that the parameters are grouped together. But I also think there are inconsistencies:

1. Other methods have a final `options?: RequestSettings` parameter, but this does not.
2. All the parameters in the `CopyDataClassPayload` are being passed in the `POST` request payload, not just the necessary parameters.
3. `dataClassId` is the source data class ID, but other properties are clearly labelled `source...`

I think to remain consistent with the rest of `mdm-resources`:

- [x] Move the `targetDataModelId`, `sourceDataModelId` properties as parameters to this method instead.
- [x] The `CopyDataClassPayload` should only contain `copyLabel` and `copyPermissions`
- [x] Add the missing optional `options:? RequestSettings` parameter

File: src/data-models/mdm-data-element.resource.ts [hunk 2]
```diff
@@ -204,9 +205,26 @@ export class MdmDataElementResource extends MdmResource {
...
+   * `HTTP POST` - Same as above but takes a payload object instead of individual parameters, allows renaming of copy.
+   *
+   * @param dataElementId The unique identifier of the data element to copy.
+   * @param copyDataElementPayload The payload object containing the source and target data model and data class identifiers.
+   *
+   * `200 OK` - will return a {@link DataElementDetailResponse} containing the new copy of a {@link DataElementDetail} object.
+   */
+  copy(
+    dataElementId: Uuid,
+    copyDataElementPayload: CopyDataElementPayload,
```
Comment (pjmonks) [inline src/data-models/mdm-data-element.resource.ts]: Same with this method:

- [x] Move `target...` and `source...` properties of the payload into parameters
- [x] `CopyDataElementPayload` should only contain `copyLabel` and `copyPermissions`

File: src/data-models/mdm-data-model.resource.ts [hunk 3]
```diff
@@ -535,4 +536,22 @@ export class MdmDataModelResource extends MdmModelDomainResource {
...
+   *
+   * @param dataModelId The unique identifier of the Data Model to copy.
+   * @param CopyContainerPayload The payload containing details on where to copy the Data Model to.
+   * @returns The result of the `PUT` request.
+   *
+   * `200 OK` - will return a {@link DataModelDetailResponse} containing a {@link DataModelDetail
+   */
+  copy(
+    dataModelId: Uuid,
+    copyContainerPayload: CopyContainerPayload
```
Comment (pjmonks) [inline src/data-models/mdm-data-model.resource.ts]: More consistency things:

- [ ] Move `target...` and `source...` properties into function parameters
- [x] `copyContainerPayload` is not the correct name for this, this implies we're copying a container (like a folder) whereas this is for a model. `CopyModelPayload` would make more sense

File: src/mdm-container-types.model.ts [hunk 4]
```diff
@@ -43,3 +43,22 @@ export interface ContainerUpdatePayload extends Payload {
...
+   */
+  folderId: Uuid;
+  /**
+   * The name of the target data model.
+   */
+  label: string;
+  /**
+   * Flag to indicate if permissions should be copied.
+   */
+  copyPermissions: Boolean;
```
Comment (pjmonks) [inline src/mdm-container-types.model.ts]: I don't think there is a `Boolean` type in TypeScript, I think this is meant to be the primitive `boolean`

File: src/terminologies/mdm-code-set.resource.ts [hunk 5]
```diff
@@ -250,4 +251,20 @@ export class MdmCodeSetResource extends MdmModelDomainResource {
...
+   * `HTTP PUT` - Copy a code set.
+   *
+   * @param codeSetId The unique identifier of the code set to copy.
+   * @param CopyContainerPayload The payload to pass to the request when copying the code set.
+   * @returns The result of the `PUT` request.
+   *
+   * `200 OK` - will return a {@link CodeSetDetailResponse} containing a {@link CodeSetDetail} object.
+   */
+  copy(codeSetId: Uuid,
+    copyContainerPayload: CopyContainerPayload
```
Comment (pjmonks) [inline src/terminologies/mdm-code-set.resource.ts]: More consistency things:

- [x] Move `target...` and `source...` properties into function parameters
- [x] `copyContainerPayload` is not the correct name for this, this implies we're copying a container (like a folder) whereas this is for a model. `CopyModelPayload` would make more sense

File: src/terminologies/mdm-term.resource.ts [hunk 6]
```diff
@@ -252,4 +255,24 @@ export class MdmTermResource extends MdmResource {
...
+    *
+    * @param terminologyId The unique identifier of the terminology to copy the term to.
+    * @param termId The unique identifier of the term to copy.
+    * @param copyTermPayload The payload of the request containing all the details for the term to copy.
+    * @returns The result of the `POST` request.
+    *
+    * `200 OK` - will return a {@link TermDetailResponse} containing a {@link TermDetail} object.
+  */
+  copy(
+    terminologyId: Uuid,
```
Comment (pjmonks) [inline src/terminologies/mdm-term.resource.ts]: More consistency things:

- [x] Move `target...` and `source...` properties into function parameters
- [x] Make it clear which IDs are the source and which are the target

File: src/terminologies/mdm-terminology.resource.ts [hunk 7]
```diff
@@ -243,4 +244,21 @@ export class MdmTerminologyResource extends MdmModelDomainResource {
...
+   *
+   * @param terminologyId The unique identifier of the terminology to copy.
+   * @param CopyContainerPayload The payload to pass to the request when copying the terminology.
+   * @returns The result of the `PUT` request.
+   *
+   * `200 OK` - will return a {@link TerminologyDetailResponse} containing a {@link TerminologyDetail} object.
+   */
+  copy(
+    terminologyId: Uuid,
+    copyContainerPayload: CopyContainerPayload
```
Comment (pjmonks) [inline src/terminologies/mdm-terminology.resource.ts]: More consistency things:

- [x] Move `target...` and `source...` properties into function parameters
- [x] Make it clear which IDs are the source and which are the target
- [x] `copyContainerPayload` is not the correct name for this, this implies we're copying a container (like a folder) whereas this is for a model. `CopyModelPayload` would make more sense

**Review by pjmonks — CHANGES_REQUESTED** [review]
Please review the inconsistencies

File: src/data-models/mdm-data-model.resource.ts (hunk 3 above)
Comment (OButlerOcc) [reply src/data-models/mdm-data-model.resource.ts]: target folder is part of the payload on this one

File: src/terminologies/mdm-code-set.resource.ts (hunk 5 above)
Comment (OButlerOcc) [reply src/terminologies/mdm-code-set.resource.ts]: targetFolder is a payload for these three

File: src/terminologies/mdm-term.resource.ts (hunk 6 above)
Comment (OButlerOcc) [reply src/terminologies/mdm-term.resource.ts]: targetTerminologyId is a payload requirement

**Review by pjmonks — APPROVED** [review]

**Pull request merged**